OpenKh.Bbs/Pmo.cs

```diff
@@ -0,0 +1,716 @@
+                            meshChunk.colors.Add(col);
+                            break;
+                    }
+
+                    Vector3 currentVertex;
+
+                    // Handle triangles and triangle strips.
+                    switch (VertexPositionFormat)
+                    {
+                        case CoordinateFormat.NORMALIZED_8_BITS:
+                            currentVertex.X = r.ReadSByte() / 127.0f;
+                            currentVertex.Y = r.ReadSByte() / 127.0f;
+                            currentVertex.Z = r.ReadSByte() / 127.0f;
+                            meshChunk.vertices.Add(currentVertex);
+                            break;
+                        case CoordinateFormat.NORMALIZED_16_BITS:
+                            vertexIncreaseAmount = ((0x2 - (Convert.ToInt32(stream.Position - vertexStartPos) & 0x1)) & 0x1);
+                            stream.Seek(vertexIncreaseAmount, SeekOrigin.Current);
+
+                            currentVertex.X = (float)stream.ReadInt16() / 32767.0f;
+                            currentVertex.Y = (float)stream.ReadInt16() / 32767.0f;
+                            currentVertex.Z = (float)stream.ReadInt16() / 32767.0f;
+                            meshChunk.vertices.Add(currentVertex);
+                            break;
+                        case CoordinateFormat.FLOAT_32_BITS:
+                            vertexIncreaseAmount = ((0x4 - (Convert.ToInt32(stream.Position - vertexStartPos) & 0x3)) & 0x3);
+                            stream.Seek(vertexIncreaseAmount, SeekOrigin.Current);
+
+                            currentVertex.X = stream.ReadFloat();
+                            currentVertex.Y = stream.ReadFloat();
+                            currentVertex.Z = stream.ReadFloat();
+                            meshChunk.vertices.Add(currentVertex);
+                            break;
+                    }
+
+                    stream.Seek(vertexStartPos + meshChunk.SectionInfo.VertexSize, SeekOrigin.Begin);
+
+                    if (flags.Primitive == PrimitiveType.PRIMITIVE_TRIANGLE)
+                    {
+                        meshChunk.Indices.Add(v);
+                    }
+                }
+
+                VertCnt = meshChunk.SectionInfo.VertexCount;
+                pmo.Meshes.Add(meshChunk);
+
+                stream.Seek(positionAfterHeader + (meshChunk.SectionInfo.VertexCount * meshChunk.SectionInfo.VertexSize), SeekOrigin.Begin);
+                stream.Seek(stream.Position % 4, SeekOrigin.Current);
+            }
+        }
+
+        public static Pmo Read(Stream stream)
+        {
+            Pmo pmo = new Pmo();
+
+            ReadHeader(ref stream, ref pmo);
+            ReadTextureSection(ref stream, ref pmo);
+
+            // Read all data
+            UInt16 VertCnt = 0xFFFF;
+            int meshSecCount = 0;
+
+            ReadMeshData(ref stream, ref pmo, ref VertCnt, ref meshSecCount, 0);
```

Comment:
I do not think you need `ref` to `stream` and `pmo`. And as you are always initialising `VertCnt` I think you can even inline that number as parameter.

Reply:
I can definitely do without ```VertCnt```, but ```stream``` and ```pmo``` need to be ```ref``` or when you read the second mesh (that uses the same method) you won't be modifying the same object.